modules/transactionpool/update.go

```diff
@@ -387,3 +388,245 @@ func (tp *TransactionPool) PurgeTransactionPool() {
+				// transaction has already been marked for removal, along with
+				// any of its output objects.
+				continue
+			}
+			// The tx index is not present in data, but i is the index where it
+			// would be inserted to maintain the sorted order. Insert the index
+			// into the slice marking removed transactions.
+			removedIndices = removedIndices[0 : len(removedIndices)+1]
+			copy(removedIndices[i+1:], removedIndices[i:])
+			removedIndices[i] = loc.txIndex
+
+			// Add all output objects to the removal stack.
+			addTxnToRemovalStack(set[loc.txIndex])
+		}
+	}
+
+	// At this point all transaction sets which have had transactions removed,
+	// will be marked in removedTxnIndices. Now we clean up each set
+	// individually, creating new transaction sets and a diff along the way.
+	diff := new(modules.TransactionPoolDiff)
+
+	for setID, txnIndices := range removedTxnIndices {
+		// Prepare changes to diff.
+		var unconfirmedSet modules.UnconfirmedTransactionSet
+		unconfirmedSet.Change = cc
+
+		numIndices := len(removedTxnIndices)
+		indexPointer := 0
+		nextIndex := txnIndices[indexPointer]
+
+		set := tp.transactionSets[setID]
+		for i := 0; i < len(set); i++ {
+			// Increment the pointer to the slice of removed transaction indices.
+			for indexPointer < numIndices-1 && nextIndex < i {
+				indexPointer++
+				if indexPointer > numIndices-1 {
+					break
+				}
+				nextIndex = txnIndices[indexPointer]
+			}
+
+			// If this transaction has been marked for removal, it can't need to
+			// be added to the new set replacing this set.
+			if i != txnIndices[indexPointer] {
+				// Remove this transaction's ID from the tpool's state.
+				delete(tp.transactionHeights, set[i].ID())
+				continue
+			}
+			// This transaction has been untouched by the removal process, so it
+			// goes back into a transaction set.
+			unconfirmedSet.IDs = append(unconfirmedSet.IDs, set[i].ID())
+			unconfirmedSet.Sizes = append(unconfirmedSet.Sizes, uint64(set[i].MarshalSiaSize()))
+			unconfirmedSet.Transactions = append(unconfirmedSet.Transactions, set[i])
+		}
+		if len(unconfirmedSet.Transactions) == 0 {
+			continue
+		}
+		// Add the changes made to this set to the diff.
+		unconfirmedSet.ID = modules.TransactionSetID(crypto.HashObject(unconfirmedSet.Transactions))
+		diff.AppliedTransactions = append(diff.AppliedTransactions, &unconfirmedSet)
+		diff.RevertedTransactions = append(diff.RevertedTransactions, modules.TransactionSetID(setID))
+
+		// Make the necessary changes to the tpool's state.
+		delete(tp.subscriberSets, setID)
+		tp.subscriberSets[TransactionSetID(unconfirmedSet.ID)] = &unconfirmedSet
+		delete(tp.transactionSetDiffs, setID)
+		tp.transactionSetDiffs[TransactionSetID(unconfirmedSet.ID)] = cc
+	}
+
+	// Delete all removed objects from tpool state.
+	for object, _ := range removedObjects {
+		delete(tp.knownObjects, object)
+	}
+
+	// Update subscribers on changes that have been made.
+	for _, subscriber := range tp.subscribers {
```

Comment:
It might be better to return a diff to whatever method calls this one (i.e. `ProcessConsensusChange`), instead of handing it off to subscribers here.